File: frontend/utils/shared-types.ts
/**
 * This file contains type defintions which are shared between the front and back end applications
 */

export type GameData = {
     turn_index: number,
     player_order: Array<string>,
}

export type Event<Code, PayloadType> = {
     event_code: Code,
     message?: string,
     data?: PayloadType,
}

export type ServerEvent = Event<ServerEventCode, ServerEventData>
export type ClientEvent = Event<ClientEventCode, ClientEventData>

export type ServerEventData = {
     session_id?: string,
     client_id?: string,
     session_client_ids?: Array<string>,
     game_data?: GameData,
}

export type ClientEventData = {
     target_ids?: Array<string>,
     session_id?: string,
     column?: number,
}

export enum ServerEventCode {
    /**
     * Session Related
     */
    ClientJoined = 1,
    ClientLeft,
    GameStarted,
    SessionResponse,
    /**
     * Game Related
     */
    TurnStart,
    LogicError,
}

export enum ClientEventCode {
    /**
     * Session Related Events
     */
    JoinSession = 1,
    CreateSession,
    LeaveSession,
    SessionRequest,
    /**
     * Game Related Events
     */
    StartGame,
    Play,
}

File: backend/server/src/shared_types.rs
/**
 * This file contains type defintions which are shared between the front and back end applications
 */
use serde::{Deserialize, Serialize};
use serde_repr::{Deserialize_repr, Serialize_repr};

#[derive(Serialize)]
pub struct ServerEventData {
    pub session_id: Option<String>,
    pub client_id: Option<String>,
    pub session_client_ids: Option<Vec<String>>,
    pub game_data: Option<GameData>,
}

#[derive(Serialize, Debug, Clone)]
pub struct GameData {
    pub turn_index: usize,
    pub player_order: Vec<String>,
}

#[derive(Serialize)]
pub struct ServerEvent {
    pub event_code: ServerEventCode,
    pub message: Option<String>,
    pub data: Option<ServerEventData>,
}

#[derive(Serialize_repr)]
#[repr(u8)]
pub enum ServerEventCode {
    // session_id, client_id, session_client_ids
    ClientJoined = 1,
    // client_id
    ClientLeft,
    GameStarted,
    // session_id, session_client_ids
    DataResponse,
    // client_id
    TurnStart,
    LogicError,
}

#[derive(Deserialize)]
pub struct ClientEvent {
    pub event_code: ClientEventCode,
    pub target_ids: Option<Vec<String>>,
    pub session_id: Option<String>,
}

#[derive(Deserialize_repr)]
#[repr(u8)]
pub enum ClientEventCode {
    // session_id
    JoinSession = 1,
    CreateSession,
    LeaveSession,
    DataRequest,
    StartGame,
    EndTurn,
    PlayCard,
    StateResponse,
}
